### PR TITLE
MM-57320 Fix autocomplete occasionally erasing all text after caret

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
I think I've finally tracked down the cause of this nasty bug where autocompleting something earlier in the post textbox will occasionally wipe out everything after the autocompleted text. If I understand correctly, the issue is because
1. https://github.com/mattermost/mattermost-webapp/pull/2078 added the ability to press enter while results are loading and then have the autocomplete wait to trigger until those results are loaded. The important part of this is that, the `SuggestionBox` passes a `resultCallback` into the `Provider` which calls `SuggestionBox.handleCompleteWord` whenever results are received.
2. Providers can call their `resultCallback` multiple times so that they can return a mix of local data and data from the server. When the above logic triggers, this means that it can call `SuggestionBox.handleCompleteWord` multiple times.
3. Calling `SuggestionBox.handleCompleteWord` multiple times should only mess up the text a bit by repeating part of the suggestion like `@userser`, but in an attempt to prevent that, https://github.com/mattermost/mattermost-webapp/pull/442 accidentally made it so that, if this triggers, the contents of the post textbox are replaced by the "pretext" which is just the text before the caret. That erases everything after the caret.

I'm still not 100% confident that this is the only issue going on, but to be safe:
1. For point 1 above, the `resultCallback` passed into the `Provider` can only trigger `SuggestionBox.handleCompleteWord` once.
2. Point 2 above is unchanged
3. For point 3, I changed the fix from the old PR to just return early if we're unable to determine what to replace with the suggestion. Ideally, we might be able to prevent this from even being possible by changing how we track the state of the textbox and what text should be replaced, but that's a big task.

@larkox I ended up reverting that previous fix you made for this issue as well because it was making some of the new tests behave funny. I've also got some other followup misc fixes to make.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57320

#### Release Note
```release-note
Fixed a longstanding bug where the at-mention autocomplete could erase post text following the autocompleted at-mention
```

### CODE REVIEW

File: webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.test.tsx [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        await userEvent.keyboard('{enter}');
+
+        await waitFor(() => {
+            expect(screen.getByPlaceholderText('test input')).toHaveValue('testtest ');
+        });
+
+        expect(screen.queryByRole('list')).not.toBeInTheDocument();
+    });
+
+    test('MM-57320 completing text with enter and calling resultCallback twice should not erase text following caret', async () => {
```
Comment (hmhealey) [inline webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.test.tsx:183]: I've been looking at this test code for far too long, so feel free to tell me if anything looks bizarre here. This component is super convoluted, and trying to cause Testing Library to trigger the bug is a mess

File: webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.jsx [hunk 2]
```diff
@@ -565,29 +568,6 @@ export default class SuggestionBox extends React.PureComponent {
...
-    };
-
-    handleKeyUp = (e) => {
-        this.confirmPretext();
-        if (this.props.onKeyUp) {
-            this.props.onKeyUp(e);
-        }
-    };
-
-    handleMouseUp = (e) => {
```
Comment (larkox) [inline webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.jsx:584]: Why we no longer need the mouse and key events?

Comment (hmhealey) [reply webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.jsx:584]: We pass `{...props}` through to the `QuickInput` below, so the props are still being passed through as they were before. The `SuggestionBox` doesn't care about them itself any more

**Review by pvev — APPROVED** [review]
LGTM! nice work @hmhealey 🚀

File: webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.jsx [hunk 3]
```diff
@@ -248,6 +248,10 @@ export default class SuggestionBox extends React.PureComponent {
         }
     }
 
+    componentWillUnmount() {
```
Comment (pvev) [inline webapp/channels/src/components/suggestion/suggestion_box/suggestion_box.jsx:251]: 👍

**Review by yasserfaraazkhan — APPROVED** [review]

**Pull request merged**